dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +253,282 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
     {
-        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Windows::PushNotifications::Feature_PushNotifications::IsEnabled());
+        {
+            auto lock{ m_lock.lock_shared() };
+            THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), m_comActivatorRegistration, "Must register event handlers before calling Register().");
+        }
 
-        THROW_HR_IF(E_INVALIDARG, activators == PushNotificationRegistrationActivators::Undefined);
+        auto lock{ m_lock.lock_exclusive() };
+        return m_foregroundHandlers.add(handler);
+    }
 
-        auto isBackgroundTaskFlagSet{ WI_IsAnyFlagSet(activators, PushNotificationRegistrationActivators::PushTrigger | PushNotificationRegistrationActivators::ComActivator) };
-        auto isProtocolActivatorSet{ WI_IsFlagSet(activators, PushNotificationRegistrationActivators::ProtocolActivator) };
 
-        THROW_HR_IF(E_INVALIDARG, isBackgroundTaskFlagSet && isProtocolActivatorSet); // Invalid flag combination
-        if (AppModel::Identity::IsPackagedProcess() && isBackgroundTaskFlagSet)
+    void PushNotificationManager::PushReceived(winrt::event_token const& token) noexcept
+    {
+        bool registeredEvent{ false };
         {
-            return true;
+            auto lock{ m_lock.lock_exclusive() };
+            m_foregroundHandlers.remove(token);
+            registeredEvent = bool(m_foregroundHandlers);
         }
-        else if(!AppModel::Identity::IsPackagedProcess() && isProtocolActivatorSet)
+
+        if (!registeredEvent)
         {
-            return true;
+            // Packaged apps with BI available will remove their handlers from the platform.
+            // Unpackaged apps / Packaged apps treated as unpackaged will unregister from Long Running process.
+            if (PushNotificationHelpers::IsPackagedAppScenario())
+            {
+                auto appUserModelId{ PushNotificationHelpers::GetAppUserModelId() };
+
+                THROW_IF_FAILED(PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get()));
+            }
+            else
+            {
+                auto notificationsLongRunningPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+
+                wil::unique_cotaskmem_string processName;
+                THROW_IF_FAILED(GetCurrentProcessPath(processName));
+
+                THROW_IF_FAILED(notificationsLongRunningPlatform->UnregisterForegroundActivator(processName.get()));
+            }
         }
-        return false;
     }
+
+    HRESULT __stdcall PushNotificationManager::InvokeAll(_In_ ULONG length, _In_ byte* payload, _Out_ BOOL* foregroundHandled) noexcept try
+    {
+        auto args = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationReceivedEventArgs>(payload, length);
+        m_foregroundHandlers(*this, args);
```

Comment:
@pmpurifoy should this have a first notification check to decide whether it needs to be sent to args or the event handler?

Reply:
Where is that logic for Push BTW? It needs to be similar to AppNotification's implementation.

Reply:
InvokeAll is only used by foreground handlers. InvokeAll isn't the entry point for the app so there doesn't need to be a first notification check. LRP will launch via protocol/cocreate if the foreground handlers fail.

Reply:
if we remove handle set the foregroundhandled = bool(foregroundhandlers)